### PR TITLE
make sure units are always stored in unit_ptr 

### PR DESCRIPTION
**Pull request by gfgtdf**

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
Why did you separate construction and initialization? You could've easily just made the existing constructors private and then added a template create function:

```
template<typename... Args>
static unit_ptr create(Args... args) {
    return unit_ptr(new unit(args...));
}
```

**Comment by gfgtdf** [comment]
> Why did you separate construction and initialization

becasue that exactly is the main point in this pr. Previously the unit ctor could invoke arbitialry umnc lua code, for example via advance_to -> add_modification -> resources::lua_kernel->apply_effect and during that call the unit was not owned by a unit_prt yet becasue it happend in  the units ctor.

**Comment by CelticMinstrel** [comment]
Ah, I see... okay. I guess that does make sense.

By the way, with this, #2615 would be fixed _but_ those abilities would be broken (using the out-of-date unit), correct?

**Comment by gfgtdf** [comment]
> By the way, with this, #2615 would be fixed but those abilities would be broken (using the out-of-date unit), correct?

well with this it be fixed if we changed attack structs back to use 'unit_ptr's instead of 'const unit*'

**Comment by gfgtdf** [comment]
ok it seems to pass now, so i'll probaly merge it soon.

**Comment by CelticMinstrel** [comment]
More complicated how?

**Comment by gfgtdf** [comment]
> More complicated how?

i thought that the specital_context_t dtor might crash if the attack_type it references begomes invalid , but that is better fixed by making specital_context_t hold a shared_ptr<attack_type> so no need for adding add unit_ptrs in the specials_context_t struct.

File: src/units/unit.hpp [hunk 1]
```diff
@@ -108,20 +108,42 @@ class unit
 
 	/** The path to the leader crown overlay. */
 	static const std::string& leader_crown();
+private:
+	void init(const config& cfg, bool use_traits = false, const vconfig* vcfg = nullptr);
+
+	void init(const unit_type& t, int side, bool real_unit, unit_race::GENDER gender = unit_race::NUM_GENDERS);
+
+	// Copy constructor
+	unit(const unit& u);
```
Comment (Vultraz) [inline src/units/unit.hpp]: Don't mark it private, leave it public but delete it.

File: src/units/unit.hpp [hunk 2]
```diff
@@ -108,20 +108,42 @@ class unit
...
+	
+	unit();
```
Comment (Vultraz) [inline src/units/unit.hpp]: Don't mark it private, leave it public but delete it.

**Comment by Vultraz** [comment]
Wait, nevermind.

**Comment by CelticMinstrel** [comment]
Merged.

**Pull request closed without merging**